Python/chapter03/p05_sort_stack/miguelHx.py

```diff
@@ -0,0 +1,274 @@
+        if it is defined, and the object is considered true if its result is nonzero.
+        If a class defines neither len() nor bool(), all its instances are considered true.
+        Returns:
+            bool: False when empty, True otherwise
+        """
+        return self._size > 0
+
+    def __len__(self) -> int:
+        return self._size
+
+    def __str__(self):
+        if self._size == 0:
+            return '<Empty>'
+        values = []
+        n = self.top
+        while n.next is not None:
+            values.append(str(n.data))
+            n = n.next
+        values.append(str(n.data))
+        return '->'.join(values)
+
+
+class TestMyStack(unittest.TestCase):
+    def test_stack_push(self):
+        s = MyStack()
+        self.assertEqual(len(s), 0)
+        self.assertEqual(s.top, None)
+        s.push(2)
+        self.assertEqual(len(s), 1)
+        self.assertEqual(s.top.data, 2)
+        self.assertEqual(s.top.next, None)
+        s.push(3)
+        self.assertEqual(len(s), 2)
+        self.assertEqual(s.top.data, 3)
+        self.assertEqual(s.top.next.data, 2)
+        s.push(4)
+        self.assertEqual(len(s), 3)
+        self.assertEqual(s.top.data, 4)
+        self.assertEqual(s.top.next.data, 3)
+        self.assertEqual(list(s), [4, 3, 2])
+
+        # test adding different types, (float and int)
+        s.push(1.2)
+        self.assertEqual(len(s), 4)
+        self.assertEqual(s.top.data, 1.2)
+        self.assertEqual(s.top.next.data, 4)
+        self.assertEqual(list(s), [1.2, 4, 3, 2])
+
+    def test_stack_peek(self):
+        s = MyStack()
+        with self.assertRaises(IndexError):
+            s.peek()
+        s.push(1)
+        s.push(2)
+        s.push(99)
+        top_val = s.peek()
+        self.assertEqual(top_val, 99)
+
+    def test_stack_pop(self):
+        # first case, attempt to pop an empty stack
+        s = MyStack()
+        with self.assertRaises(IndexError):
+            s.pop()
+        s.push(1)
+        s.push(2)
+        s.push(3)
+        # size is 3
+        self.assertEqual(list(s), [3, 2, 1])
+        val = s.pop()
+        self.assertEqual(val, 3)
+        self.assertEqual(s._size, 2) # size should now be 2
+        self.assertEqual(list(s), [2, 1])
+
+    def test__bool__(self):
+        s = MyStack()
+        self.assertFalse(s)
+        s.push(3)
+        self.assertTrue(s)
+
+
+def sorted_stack(stack: MyStack[T]) -> MyStack[T]:
+    """This function will take in a stack
+    and return a sorted copy.
+    In order to do this, we need to find the biggest
+    number in an input stack copy, and then push it to the
+    output stack. Repeat until input stack copy is empty.
+
+    Args:
+        stack (MyStack): stack of items
+
+    Returns:
+        MyStack: sorted stack of items, with smallest items on top
+    """
+    # create empty output stack
+    output_stack: MyStack = MyStack()
+    # make copy of input stack.
+    stack_copy: MyStack = stack.produce_copy()
```

Comment:
it may be fine to make this operation destructive to avoid having to copy the original stack (although copying a stack could be achieved only through stack operations).

Reply:
Operation made destructive in re-implementation here:  ebee21966031a34637f2b47c0eaabf9d96ca8d5d